src/qrc11.asm

```diff
@@ -0,0 +1,359 @@
+; https://github.com/leiradel/qrc1
+
+; ------------------------------------------------------------------------
+; Macros for some undocumented Z80 instructions.
+; ------------------------------------------------------------------------
+
+addixl: macro
+    db $dd
+    add a, l
+endm
+
+ldixh_a: macro
+    db $dd
+    ld h, a
+endm
+
+decixh: macro
+    db $dd
+    dec h
+endm
+
+decixl: macro
+    db $dd
+    dec l
+endm
+
+; qrc11_message contains a $40 byte followed by the message length followed
+; by the message (maximum 251 bytes).
+qrc11_encmessage:
+    ; ------------------------------------------------------------------------
+    ; Encode the message.
+    ; ------------------------------------------------------------------------
+
+    ; Insert a 0000 nibble to before the length
+    ld hl, qrc11_message + 1
+    ld c, (hl)
+    xor a
+
+    ; Shift the message to the right by four bits.
+    ld b, c
+    inc b
+
+qrc11_shift_msg:
+        rrd
+        inc hl
+    djnz qrc11_shift_msg
+
+    ; A has the low nibble of the last message byte, shift it to the high
+    ; nibble and set the low nibble to 0, which is the end of message mark.
+    ld (hl), 0
+    rrd
+    inc hl
+
+    ; Pad the rest of the message with $ec and $11.
+    ld a, 251
+    sub c
+    jr z, qrc11_no_padding
+
+    ld b, a
+    ld a, $ec
+qrc11_pad_msg:
+        ld (hl), a
+        inc hl
+        xor $fd
+    djnz qrc11_pad_msg
+
+qrc11_no_padding:
+
+    ; ------------------------------------------------------------------------
+    ; Calculate the message ECC.
+    ; ------------------------------------------------------------------------
+
+    ; Copy each block of the original encoded message to the target buffer,
```

Comment:
I think this comment doesn't hold anymore, `qrc11_ecc` will save and restore the message block.